src/wrapper/errors.rs

```diff
@@ -1,84 +1,74 @@
-        NullDeref(context: &'static str) {
-            description("null pointer deref")
-            display("null pointer deref in {}", context)
-        }
-        TryLock {
-            description("mutex already locked")
-            display("mutex already locked")
-        }
-        JavaVMMethodNotFound(name: &'static str) {
-            description("Method pointer null in JavaVM")
-            display("JavaVM null method pointer for {}", name)
-        }
-        ThreadDetached {
-            description("Current thread is not attached to the java VM")
-            display("Current thread is not attached to the java VM")
-        }
-        Other(error: sys::jint) {
-            description("JNI error")
-            display("JNI error: {}", error)
-        }
-    }
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Invalid JValue type cast: {0}. Actual type: {1}")]
+    WrongJValueType(&'static str, &'static str),
+    #[error("Invalid constructor return type (must be void)")]
+    InvalidCtorReturn,
+    #[error("Invalid number of arguments passed to java method: {0}")]
+    InvalidArgList(TypeSignature),
+    #[error("Method not found: {name} {sig}")]
+    MethodNotFound { name: String, sig: String },
+    #[error("Field not found: {name} {sig}")]
+    FieldNotFound { name: String, sig: String },
+    #[error("Java exception was thrown")]
+    JavaException,
+    #[error("JNIEnv null method pointer for {0}")]
+    JNIEnvMethodNotFound(&'static str),
+    #[error("Null pointer in {0}")]
+    NullPtr(&'static str),
+    #[error("Null pointer deref in {0}")]
+    NullDeref(&'static str),
+    #[error("Mutex already locked")]
+    TryLock,
+    #[error("JavaVM null method pointer for {0}")]
+    JavaVMMethodNotFound(&'static str),
+    #[error("Unknown error")]
+    Unknown,
+    #[error("Current thread is not attached to the Java VM")]
+    ThreadDetached,
+    #[error("Problem due to JNI version mismatch")]
+    WrongVersion,
+    #[error("Not enough memory available")]
+    NoMemory,
+    #[error("Java VM already created")]
+    AlreadyCreated,
+    #[error("Invalid arguments provided")]
+    InvalidArguments,
+    #[error("Field already set: {0}")]
+    FieldAlreadySet(String),
+    #[error("Throw failed with error code {0}")]
+    ThrowFailed(i32),
+    #[error("Parse failed for input: {1}")]
+    ParseFailed(#[source] combine::error::StringStreamError, String),
+    #[error("JNI error: {0}")]
+    Other(sys::jint),
 }
 
-unsafe impl Sync for Error {}
-
 impl<T> From<::std::sync::TryLockError<T>> for Error {
     fn from(_: ::std::sync::TryLockError<T>) -> Self {
-        ErrorKind::TryLock.into()
+        Error::TryLock
     }
 }
 
 pub fn jni_error_code_to_result(code: sys::jint) -> Result<()> {
     match code {
         sys::JNI_OK => Ok(()),
-        sys::JNI_EDETACHED => Err(Error::from(ErrorKind::ThreadDetached)),
-        _ => Err(Error::from(ErrorKind::Other(code))),
+        sys::JNI_ERR => Err(Error::Unknown),
```

Comment:
I'm not that happy with this naming neither 🤔 . I found some really short descriptions of the errors [HERE](https://docs.oracle.com/javase/9/docs/specs/jni/functions.html) and used them for the naming of this new errors.

I personally feel that it's a little unfortunate to have this JNI error code related errors all mixed with the rest. What do you think about the idea of having a second error enum for this errors and use them as value for a single error. Something like this:

```rust
#[derive(Debug, Error)]
pub enum Error {
    ... // Other already existing errors

    #[error("JNI call failed with an error code: {0}")]
    JniCall(#[source] JniError),
}

#[derive(Debug, Error)]
pub enum JniError {
    #[error("Unknown error")]
    Unknown,
    #[error("Thread was detached from the VM")]
    Detached,
    #[error("JNI version error")]
    Version,
    #[error("Not enough memory")]
    NoMemory,
    #[error("VM already created")]
    Exists,
    #[error("Invalid arguments")]
    Invalid,
}
```

That way it would become a second layer of errors and would be more similar to other approaches where you for example have a `std::io::Error` that you wrap with `thiserror` like `IoError(#[from] std::io::Error)`. It would be clearer that this is not an actual error coming from this crate but being an error directly coming from the underlying Java JNI library.

Reply:
I like this idea, but let's wait for an opinion from @dmitry-timofeev. He is more experienced with JNI (and Java), so I would like to hear his opinion before asking you to make these changes.

Reply:
Do I understand correctly that the suggestion is to separate JNI error return code ("Return codes" in the specification and the header) from some other errors that might occur in this crate? That seems to be a good idea to me (though, I can't comment on the effects of such change in terms of ergonomics in Rust — if that remains just as good, then I'd go for it). 

We may do that in a separate or the same PR, but if we decide to make this change, it would be better to have it all in one release (to not force people to make two migrations).

Reply:
Yes, we would have a separate _sub_-error just for the JNI error return codes. That would be wrapped by the main error and behave the same way as having an inner error in error-chain. So from the caller site you can do things like:
```rust
if let Err(Error::JniCall(JniError::NoMemory))) = some_call() {
    eprintln!("We're out of memory!");
}
```

Or use it in other pattern matchings and so on. And if the error is wrapped with for example `anyhow` and is printed verbose, then we can see the error in the output:

```rust
println!("{}", anyhow::anyhow!(Error::JniCall(JniError::NoMemory)));
// Output:
// JNI call failed

println!("{:?}", anyhow::anyhow!(Error::JniCall(JniError::NoMemory)));
// Output:
// JNI call failed
//
// Caused by:
//     Not enough memory

println!("{:#?}", anyhow::anyhow!(Error::JniCall(JniError::NoMemory)));
// Output:
// JniCall(
//     NoMemory,
// )
```

For this outputs I changed the above macro slightly to just `#[error("JNI call failed")]` which is more compact and doesn't repeat the sub-error in the main message.

So I believe Rust ergonomics should be good.

I already made the changes locally to prepare these output samples so I can include it in this PR without further efforts.